### PR TITLE
KAFKA-3502: move RocksDB options construction to init()

### PR DESCRIPTION
**Pull request by guozhangwang**
In RocksDBStore, options / wOptions / fOptions are constructed in the constructor, which needs to be dismissed in the close() call; however in some tests, the generated topology is not initialized at all, and hence the corresponding state stores are supposed to not be able to be closed as well since their `init` function is not called. This could cause the above option objects to be not released.

This is fixed in this patch to move the logic out of constructor and inside `init` functions, so that no RocksDB objects will be created in the constructor only. Also some minor cleanups:

1. In KStreamTestDriver.close(), we lost the logic to close the state stores but only call `flush`; it is now changed back to call both.
2. Moved the forwarding logic from KStreamTestDriver to MockProcessorContext to remove the mutual dependency: these functions should really be in ProcessorContext, not the test driver.

### CODE REVIEW

**Comment by guozhangwang** [comment]
ping @enothereska @dguy @mjsax for reviews.

**Review by mjsax — APPROVED** [review]
LGTM

**Comment by mjsax** [comment]
This fix seems to be incomplete:
> org.apache.kafka.streams.state.internals.SegmentIteratorTest > shouldOnlyIterateOverSegmentsInRange PASSED
pure virtual method called
terminate called without an active exception

**Review by mjsax — COMMENTED** [review]
This fix seems to be incomplete:
> org.apache.kafka.streams.state.internals.SegmentIteratorTest > shouldOnlyIterateOverSegmentsInRange PASSED
pure virtual method called
terminate called without an active exception

**Comment by guozhangwang** [comment]
> This fix seems to be incomplete:

Looking into it.

**Comment by guozhangwang** [comment]
Found another issue with `SegmentIteratorTest`, where the test could be opening the underlying `Segment`'s iterator without closing it at the end if it is not calling `hasNext` until it returns false.

File: streams/src/test/java/org/apache/kafka/streams/kstream/KStreamBuilderTest.java [hunk 1]
```diff
@@ -41,7 +42,16 @@
 
 public class KStreamBuilderTest {
 
+    private static final String APP_ID = "app-id";
+
     private KStreamTestDriver driver = null;
+    private KStreamBuilder builder = null;
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/kstream/KStreamBuilderTest.java]: The `KStreamBuilder` could be created here. Can also be `final`

**Review by dguy — COMMENTED** [review]
1 minor comment, otherwise LGTM

**Review by dguy — APPROVED** [review]

**Review by mjsax — APPROVED** [review]
LGTM

**Comment by guozhangwang** [comment]
retest this please

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java [hunk 2]
```diff
@@ -62,8 +62,8 @@ public void before() {
                                                     new SessionKeySchema());
 
         stateDir = TestUtils.tempDirectory();
-        final MockProcessorContext context = new MockProcessorContext(null,
-                                                                      stateDir,
+        final MockProcessorContext context = new MockProcessorContext(
+                stateDir,
```
Comment (hachikuji) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java]: nit: the alignment got messed up on a bunch of these.

File: streams/src/test/java/org/apache/kafka/test/KStreamTestDriver.java [hunk 3]
```diff
@@ -106,48 +101,49 @@ private void initTopology(final ProcessorTopology topology, final List<StateStor
         }
     }
 
+    public ProcessorTopology topology() {
+        return topology;
+    }
+
+    public ProcessorTopology globalTopology() {
```
Comment (hachikuji) [inline streams/src/test/java/org/apache/kafka/test/KStreamTestDriver.java]: nit: seems this is unused?

**Comment by hachikuji** [comment]
@guozhangwang Other than a couple nitpicks, LGTM.

**Comment by guozhangwang** [comment]
@hachikuji addressed, would you mind taking another look?

**Comment by hachikuji** [comment]
Thanks for the patch! Fixed one more alignment nit when merging.

**Pull request closed without merging**